rfd/0141-tsh-bench-db.md

```diff
@@ -0,0 +1,279 @@
+---------- -----------------
+25         123 ms
+50         125 ms
+75         129 ms
+90         130 ms
+95         130 ms
+99         130 ms
+100        130 ms
+```
+
+### [`mysqlslap`](https://dev.mysql.com/doc/refman/5.7/en/mysqlslap.html)
+
+> `mysqlslap` is a diagnostic program designed to emulate client load for a
+MySQL server and to report the timing of each stage. It works as if multiple
+clients are accessing the server.
+
+It has to working formats:
+* Auto-generates database tables and queries. When used like this, it has flags
+  to customize table columns (set how many columns there will be) and choose the
+  type of queries it will execute during the benchmark (for example, `read` will
+  only execute SELECT queries).
+* Custom queries. It has a parameter to setup a table and insert necessary data.
+  And executes the custom query provided by the user.
+
+## Details
+The database subcommands will be divided per protocol. This gives more
+flexibility for each protocol to require their arguments and flags.
+
+The "root" command for each protocol will execute the following:
+* Starts a connection with the database (using local proxy);
+* Issue a ping-like query to ensure the connection is working. This also covers
+  drivers that connect only when queries are executed.
+* Closes the connection.
+
+User certificate generation and local proxy start will not affect the final
+results. The connections will be issued to a single local proxy, reducing the
+resources used locally.
+
+The metrics generated by those subcommands will be restricted to raw end-to-end
+timings (without any breakdown). It must be used alongside backend metrics to
+have full detailed information on the service behavior.
+
+```mermaid
+sequenceDiagram
+  participant tsh
+  participant Teleport
+  participant DB as Database
+
+  tsh->>Teleport: Generate Certificates (tsh db login)
+
+  loop every execution
+    tsh->>Teleport: Establish connection
+    Teleport->>DB: Connect
+    DB->>tsh: Connection
+    tsh->>DB: Ping
+  end
+```
+
+### Commands
+Both PostgreSQL and MySQL will have the same command arguments and flags.
+
+```code
+tsh bench postgres [--db-user=] [--db-name=] [--uri] [database]
+```
+
+```code
+tsh bench mysql [--db-user=] [--db-name=] [--uri] [database]
+```
+
+| Flag        | Description |
+| ----------- | ----------- |
+| `--db-user` | Database user used to connect to the target database. The user must have enough permissions on the database to execute all the benchmark queries. |
+| `--db-name` | Database name where benchmark queries will be executed. Depending on the database protocol, this isn't required. |
+| `--uri`     | Direct database access URI. When provided, the benchmark will issue connections directly to this database, and no Teleport is involved in the testing. It must contain all the connection information, including authentication credentials. The contents from --db-user and --db-name will be ignored when provided. |
```

Comment:
Instead of this flag, I wonder if we should just teach the command to recognize whether `database` argument is a Teleport database name or a database URI like postgres://.... Otherwise the UX is a little confusing, you need to specify database when connecting through Teleport and the flag when connecting directly.